gateway/admin.go

```diff
@@ -21,3 +32,169 @@ func (g *GatewayAdmin) Status() interface{} {
 	}
 
 }
+
+type GatewayRPCHandler struct {
+	jobsDB jobsdb.JobsDB
+}
+
+type SqlRunner struct {
+	dbHandle     *sql.DB
+	jobTableName string
+}
+
+type SourceEvents struct {
+	Count int
+	Name  string
+	ID    string
+}
+
+func (r *SqlRunner) getUniqueSources() ([]SourceEvents, error) {
+	var rows *sql.Rows
+	var err error
+	sources := make([]SourceEvents, 0)
+	uniqueSourceValsStmt := fmt.Sprintf(`select count(*) as count, parameters -> 'source_id' as source from %s group by parameters -> 'source_id';`, r.jobTableName)
+	rows, err = r.dbHandle.Query(uniqueSourceValsStmt)
+	if err != nil {
+		return sources, err
+	}
+	defer rows.Close()
+	sourceEvent := SourceEvents{}
+	for rows.Next() {
+		err = rows.Scan(&sourceEvent.Count, &sourceEvent.ID)
+		if err != nil {
+			return sources, err // defer closing of rows, so return will not memory leak
+		}
+		sources = append(sources, sourceEvent)
+	}
+
+	if err = rows.Err(); err != nil {
+		return sources, err
+	}
+
+	return sources, err
+}
+
+func (r *SqlRunner) getNumUniqueUsers() (int, error) {
+	var numUsers int
+	var err error
+	numUniqueUsersStmt := fmt.Sprintf(`select count(*) from (select  distinct user_id from %s) as t`, r.jobTableName)
+	err = runSQL(r, numUniqueUsersStmt, &numUsers)
+	return numUsers, err
+}
+
+func (r *SqlRunner) getAvgBatchSize() (float64, error) {
+	var batchSize sql.NullFloat64
+	var avgBatchSize float64
+	var err error
+	avgBatchSizeStmt := fmt.Sprintf(`select avg(jsonb_array_length(batch)) from (select event_payload->'batch' as batch from %s) t`, r.jobTableName)
+	err = runSQL(r, avgBatchSizeStmt, &batchSize)
+	if batchSize.Valid {
```

Comment:
Handle err. This can cause nil pointer exception;

Reply:
Just checked, batchSize.Valid will be true only when batchSize is NOT NULL, false when NULL. Hence wont cause a nil pointer exception